File: packages/xpute/runtime/src/ipc/change_waker.ts
// @xpute/runtime/ipc/change_waker.ts

/**
 * Push delivery over a change log: listeners per (topic, key), woken when a
 * pump finds their key in the journal.
 *
 * An adapter, not the protocol. The log never calls anyone; whoever owns the
 * pump decides when listeners run — once a tick, say — so a burst of changes
 * to one key wakes its listeners once, and a source's write never runs
 * consumer code inside it. A pump that finds its cursor overwritten wakes
 * every listener it holds: each of them re-reads, which is what `lost`
 * means.
 *
 * A topic listener hears every changed key of its topic instead, once per
 * key per pump, and null when the pump was lost — its filter is its own.
 */

import type { u32, u64 } from "@xpute/core/abi/word.ts";
import type { U32Array } from "@xpute/core/abi/array.ts";
import { Vector } from "@xpute/core/collection/arena.ts";
import { I64_KEYS, IndexedMap } from "@xpute/core/collection/map.ts";

import type { ChangeJournal, Cursor, Topic } from "./change.ts";

type Listener = () => void;
/** A changed key, or null: the pump lost its place, re-read the topic. */
export type TopicListener = (key: u64 | null) => void;

export class ChangeWaker {
  private cursor: Cursor;
  private pumps: u32 = 0;
  /** Per topic: the keys listened to, their listeners, and the pump that
   * last woke each, so a key bumped twice in one span wakes once. */
  private readonly keys: IndexedMap<u64>[] = [];
  private readonly lists: Listener[][][] = [];
  private readonly woken: Vector<U32Array>[] = [];
  /** Per topic: whole-topic listeners, and the keys already given to them
   * this pump. */
  private readonly topic_lists: TopicListener[][] = [];
  private readonly topic_seen: IndexedMap<u64>[] = [];

  constructor(private readonly log: ChangeJournal) {
    this.cursor = log.head;
    for (let t = 0; t < log.topics; t++) {
      this.keys.push(new IndexedMap(I64_KEYS, 16));
      this.lists.push([]);
      this.woken.push(new Vector(Uint32Array, { init_cap: 16 }));
      this.topic_lists.push([]);
      this.topic_seen.push(new IndexedMap(I64_KEYS, 64));
    }
  }

  subscribe(topic: Topic, key: u64, listener: Listener): () => void {
    const keys = this.keys[topic], lists = this.lists[topic], woken = this.woken[topic];
    const pos = keys.insert(key);
    if (pos === woken.len) woken.push(0);
    (lists[pos] ??= []).push(listener);
    return () => {
      const p = keys.find(key);
      if (p < 0) return;
      const list = lists[p];
      const at = list.indexOf(listener);
      if (at >= 0) list.splice(at, 1);
      if (list.length > 0) return;
      const from = keys.remove_at(p);
      if (from !== p) {
        lists[p] = lists[from];
        woken.set(p, woken.get(from));
      }
      lists.length = from;
      woken.truncate(from);
    };
  }

  /** Hears every key of `topic` that changes. */
  subscribe_topic(topic: Topic, listener: TopicListener): () => void {
    const list = this.topic_lists[topic];
    list.push(listener);
    return () => {
      const at = list.indexOf(listener);
      if (at >= 0) list.splice(at, 1);
    };
  }

  /** Wakes the listeners of every key changed since the last pump. Returns
   * how many listeners ran. */
  pump(): u32 {
    const log = this.log;
    if (this.cursor === log.head) return 0;
    const pump = ++this.pumps;
    let ran = 0;
    if (!log.readable(this.cursor)) {
      for (let t = 0; t < this.lists.length; t++) {
        for (let p = 0; p < this.keys[t].size; p++) ran += this.wake(t, p, pump);
        for (const listener of this.topic_lists[t].slice()) listener(null), ran++;
      }
    } else {
      for (const seen of this.topic_seen) seen.clear();
      for (let c = this.cursor; c < log.head; c++) {
        const i = log.slot(c);
        const t = log.topic[i];
        const key = log.key[i];
        const p = this.keys[t].find(key);
        if (p >= 0) ran += this.wake(t, p, pump);
        const topic_list = this.topic_lists[t];
        if (topic_list.length === 0) continue;
        const seen = this.topic_seen[t];
        if (seen.find(key) >= 0) continue;
        seen.insert(key);
        for (const listener of topic_list.slice()) listener(key), ran++;
      }
    }
    this.cursor = log.head;
    return ran;
  }

  private wake(topic: Topic, pos: u32, pump: u32): u32 {
    const woken = this.woken[topic];
    if (woken.get(pos) === pump) return 0;
    woken.set(pos, pump);
    // A listener may unsubscribe as it runs; walk what was there when the
    // wake began.
    const list = this.lists[topic][pos].slice();
    for (const listener of list) listener();
    return list.length;
  }
}

File: crates/xpute/runtime/src/ipc/change_waker.rs
// xpute-runtime/ipc/change_waker.rs

//! Push delivery over a change log: listeners per (topic, key), woken when a
//! pump finds their key in the journal.
//!
//! An adapter, not the protocol. The log never calls anyone; whoever owns the
//! pump decides when listeners run — once a tick, say — so a burst of changes
//! to one key wakes its listeners once, and a source's write never runs
//! consumer code inside it. A pump that finds its cursor overwritten wakes
//! every listener it holds: each of them re-reads, which is what `lost`
//! means.
//!
//! A topic listener hears every changed key of its topic instead, once per
//! key per pump, and none when the pump was lost — its filter is its own.
//!
//! `subscribe` returns a `Subscription` that `unsubscribe` takes, since a
//! closure could not hold the waker mutably. The waker holds the journal
//! with whoever writes it.

use core::cell::RefCell;
use std::collections::HashSet;
use std::rc::Rc;

use fnv::FnvBuildHasher;
use indexmap::IndexSet;

use super::change::{ChangeJournal, Cursor, Topic};

type Listener = Box<dyn FnMut()>;
/// A changed key, or none: the pump lost its place, re-read the topic.
pub type TopicListener = Box<dyn FnMut(Option<i64>)>;

/// What `subscribe` hands back, for `unsubscribe`.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct Subscription {
    topic: Topic,
    key: Option<i64>,
    id: u32,
}

pub struct ChangeWaker<J: ChangeJournal> {
    log: Rc<RefCell<J>>,
    cursor: Cursor,
    pumps: u32,
    /// Per topic: the keys listened to, their listeners, and the pump that
    /// last woke each, so a key bumped twice in one span wakes once.
    keys: Vec<IndexSet<i64, FnvBuildHasher>>,
    lists: Vec<Vec<Vec<(u32, Listener)>>>,
    woken: Vec<Vec<u32>>,
    /// Per topic: whole-topic listeners, and the keys already given to them
    /// this pump.
    topic_lists: Vec<Vec<(u32, TopicListener)>>,
    topic_seen: Vec<HashSet<i64, FnvBuildHasher>>,
    next_id: u32,
}

impl<J: ChangeJournal> ChangeWaker<J> {
    pub fn new(log: Rc<RefCell<J>>) -> ChangeWaker<J> {
        let (cursor, topics) = {
            let l = log.borrow();
            (l.head(), l.topics())
        };
        let mut waker = ChangeWaker {
            log,
            cursor,
            pumps: 0,
            keys: Vec::new(),
            lists: Vec::new(),
            woken: Vec::new(),
            topic_lists: Vec::new(),
            topic_seen: Vec::new(),
            next_id: 0,
        };
        for _ in 0..topics {
            waker.keys.push(IndexSet::with_capacity_and_hasher(16, FnvBuildHasher::default()));
            waker.lists.push(Vec::new());
            waker.woken.push(Vec::with_capacity(16));
            waker.topic_lists.push(Vec::new());
            waker.topic_seen.push(HashSet::with_capacity_and_hasher(64, FnvBuildHasher::default()));
        }
        waker
    }

    pub fn subscribe(&mut self, topic: Topic, key: i64, listener: Listener) -> Subscription {
        let t = topic as usize;
        let (pos, _) = self.keys[t].insert_full(key);
        if pos == self.woken[t].len() {
            self.woken[t].push(0);
        }
        if self.lists[t].len() <= pos {
            self.lists[t].resize_with(pos + 1, Vec::new);
        }
        let id = self.next_id;
        self.next_id += 1;
        self.lists[t][pos].push((id, listener));
        Subscription { topic, key: Some(key), id }
    }

    /// Hears every key of `topic` that changes.
    pub fn subscribe_topic(&mut self, topic: Topic, listener: TopicListener) -> Subscription {
        let id = self.next_id;
        self.next_id += 1;
        self.topic_lists[topic as usize].push((id, listener));
        Subscription { topic, key: None, id }
    }

    /// The closure `subscribe` returns there.
    pub fn unsubscribe(&mut self, sub: Subscription) {
        let t = sub.topic as usize;
        let Some(key) = sub.key else {
            self.topic_lists[t].retain(|(id, _)| *id != sub.id);
            return;
        };
        let Some(p) = self.keys[t].get_index_of(&key) else { return };
        self.lists[t][p].retain(|(id, _)| *id != sub.id);
        if !self.lists[t][p].is_empty() {
            return;
        }
        // The last position moves into the hole; the lanes follow it.
        let from = self.keys[t].len() - 1;
        self.keys[t].swap_remove_index(p);
        if from != p {
            let moved = self.lists[t].swap_remove(from);
            self.lists[t][p] = moved;
            self.woken[t][p] = self.woken[t][from];
        } else {
            self.lists[t].truncate(from);
        }
        self.woken[t].truncate(from);
    }

    /// Wakes the listeners of every key changed since the last pump. Returns
    /// how many listeners ran.
    pub fn pump(&mut self) -> u32 {
        let head = self.log.borrow().head();
        if self.cursor == head {
            return 0;
        }
        self.pumps += 1;
        let pump = self.pumps;
        let mut ran = 0;
        let readable = self.log.borrow().readable(self.cursor);
        if !readable {
            for t in 0..self.lists.len() {
                for p in 0..self.keys[t].len() {
                    ran += self.wake(t as Topic, p as u32, pump);
                }
                for (_, listener) in self.topic_lists[t].iter_mut() {
                    listener(None);
                    ran += 1;
                }
            }
        } else {
            for seen in self.topic_seen.iter_mut() {
                seen.clear();
            }
            for c in self.cursor..head {
                let (t, key) = {
                    let log = self.log.borrow();
                    let i = log.slot(c) as usize;
                    (log.topic()[i] as usize, log.key()[i])
                };
                if let Some(p) = self.keys[t].get_index_of(&key) {
                    ran += self.wake(t as Topic, p as u32, pump);
                }
                if self.topic_lists[t].is_empty() || !self.topic_seen[t].insert(key) {
                    continue;
                }
                for (_, listener) in self.topic_lists[t].iter_mut() {
                    listener(Some(key));
                    ran += 1;
                }
            }
        }
        self.cursor = head;
        ran
    }

    fn wake(&mut self, topic: Topic, pos: u32, pump: u32) -> u32 {
        let woken = &mut self.woken[topic as usize];
        if woken[pos as usize] == pump {
            return 0;
        }
        woken[pos as usize] = pump;
        // The list is borrowed for the walk, so a listener cannot unsubscribe
        // as it runs: an unsubscribe is a call on the waker after it.
        let list = &mut self.lists[topic as usize][pos as usize];
        for (_, listener) in list.iter_mut() {
            listener();
        }
        list.len() as u32
    }
}
